### PR TITLE
Net-2713/resource apply command

### PR DESCRIPTION
**Pull request by JadhavPoonam**
### Description

- Add support for the `apply` command to create/update a resource.
- Updated dependencies in `api`, `envoy-extensions`, and `troubleshoot` sub modules.

Usage:
`consul resource apply -f=<file-path>`

### Testing & Reproduction steps

- Add unit tests for validations and correctness
- Manually tested the command in both server and client mode

### PR Checklist

* [ ] ~updated test coverage~
* [ ] ~external facing docs updated~
* [x] appropriate backport labels added
* [ ] ~not a security concern~

### CODE REVIEW

File: command/resource/apply/apply.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+func (c *cmd) Help() string {
+	return flags.Usage(c.help, nil)
+}
+
+const synopsis = "Writes/updates resource information"
+const help = `
+Usage: consul resource apply -f=<file-path>
+
+Write and/or update a resource by providing the definition in the hcl file in the argument
```
Comment (dhiaayachi) [inline command/resource/apply/apply.go:166]:
```suggestion
Write and/or update a resource by providing the definition in an hcl file as an argument
```

File: command/resource/apply/apply.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+Example:
+
+$ consul resource apply -f=demo.hcl
+
+Sample demo.hcl:
+
+ID {
+	Type = gvk("group.version.kind")
```
Comment (dhiaayachi) [inline command/resource/apply/apply.go:175]: The gvk() magic need to be documented, probably not in the scope of this PR.

File: command/resource/apply/apply_test.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+			expectedCode: 1,
+			expectedErr:  errors.New("Flag -f is required"),
+		},
+		"file parsing failure": {
+			args:         []string{"-f=../testdata/invalid.hcl"},
+			expectedCode: 1,
+			expectedErr:  errors.New("Failed to decode resource from input file"),
+		},
+	}
+
```
Comment (dhiaayachi) [inline command/resource/apply/apply_test.go:81]: We can maybe add a test with a missing file?

**Review by dhiaayachi — COMMENTED** [review]
Thank you for this @JadhavPoonam! I left a few comments for your consideration.

File: command/resource/apply/apply_test.go (hunk 3 above)
Comment (JadhavPoonam) [reply command/resource/apply/apply_test.go:81]: The missing file is the "no file path" test case. Anything else you had in mind?

File: command/resource/apply/apply.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	c.help = flags.Usage(help, c.flags)
+}
+
+func makeWriteRequest(parsedResource *pbresource.Resource) (payload *api.WriteRequest, error error) {
+	data, err := protojson.Marshal(parsedResource.Data)
+	if err != nil {
+		return nil, fmt.Errorf("unrecognized hcl format: %s", err)
+	}
+
+	var resourceData map[string]string
```
[the remark this answers is not in the document]
Comment (JadhavPoonam) [reply command/resource/apply/apply.go]: Good catch! We can't assume this to be a `string`. Updated and added a test case for it.

File: api/resource.go [hunk 5]
```diff
@@ -42,3 +50,29 @@ func (resource *Resource) Read(gvk *GVK, resourceName string, q *QueryOptions) (
 
 	return out, nil
 }
+
+func (resource *Resource) Apply(gvk *GVK, resourceName string, q *QueryOptions, payload *WriteRequest) (map[string]interface{}, *WriteMeta, error) {
```
[the remark this answers is not in the document]
Comment (JadhavPoonam) [reply api/resource.go]: I don't think we have a usable type in the package that we can use directly here but I created a concrete type using the defined structs

File: command/resource/apply/apply_test.go (hunk 3 above)
Comment (dhiaayachi) [reply command/resource/apply/apply_test.go:89]: not sure the "no file path" test case would exercise the else branch in line 93. I was thinking about `-f /file/noexist`

**Review by dhiaayachi — APPROVED** [review]
LGTM! 🚀

File: command/resource/apply/apply_test.go (hunk 3 above)
Comment (JadhavPoonam) [reply command/resource/apply/apply_test.go:89]: Ahh... yes. I will add that!

**Pull request merged**